In [38]:
import ipeadatapy as idpy

# Basic Usage Overview

<img src="fluxogram.png" />

### Plans for this notebook:

1. **BEGINNER: Exploring Available Time Series**
    - 1.1. Having an Overview of a Time Series
    - 1.2. Checking Time Series' Data

2. **INTERMEDIATE: Exploring Ipeadata Metadata**
    - 2.1. Advanced Filtering Using Metadata

3. **Extracting Data**

4. **ADVANCED: Doing from Scratch**
 

# 1. BEGINNER: Exploring Available Time Series

To start, let's list all Ipeadata's time series:

In [3]:
idpy.list_series()

CODE                                               NAME
0          ABATE_ABPEAV            Abate - aves - peso das carcaças - qde.
1          ABATE_ABPEBV         Abate - bovinos - peso das carcaças - qde.
2          ABATE_ABPESU          Abate - suínos - peso das carcaças - qde.
3          ABATE_ABQUAV                                Abate - aves - qde.
4          ABATE_ABQUBV                             Abate - bovinos - qde.
5          ABATE_ABQUSU                              Abate - suínos - qde.
6      ABATE12_ABPEAV12            Abate - aves - peso das carcaças - qde.
7      ABATE12_ABPEBO12            Abate - bois - peso das carcaças - qde.
8      ABATE12_ABPEBV12         Abate - bovinos - peso das carcaças - qde.
9      ABATE12_ABPEFR12         Abate - frangos - peso das carcaças - qde.
10     ABATE12_ABPENO12        Abate - novilhos - peso das carcaças - qde.
11     ABATE12_ABPESU12          Abate - suínos - peso das carcaças - qde.
12     ABATE12_ABPEVA12           Abate - vacas - peso das carcaças - qde.
13     ABATE12_ABPEVI12         Abate - vitelos - peso das carcaças - qde.
14     ABATE12_ABQUAV12                                Abate - aves - qde.
15     ABATE12_ABQUBO12                                Abate - bois - qde.
16     ABATE12_ABQUBV12                             Abate - bovinos - qde.
17     ABATE12_ABQUFR12                             Abate - frangos - qde.
18     ABATE12_ABQUNO12                            Abate - novilhos - qde.
19     ABATE12_ABQUSU12                              Abate - suínos - qde.
20     ABATE12_ABQUVA12                               Abate - vacas - qde.
21     ABATE12_ABQUVI12                             Abate - vitelos - qde.
22        ABIA12_ALIM12  Produção física - alimentos - índice (média 19...
23       ABPO12_PAPEL12  Expedição de caixas, acessórios e chapas - pap...
24    ABRACAL_PCALCARIO                Produção - calcário agrícola - qde.
25      ANATEL_APARELHO                    Telefone - aparelhos instalados
26          ANATEL_INST                Telefone - acessos fixos instalados
27          ANATEL_SERV                Telefone - acessos fixos em serviço
28       ANATEL_SERVMOV               Telefone - acessos móveis em serviço
29    ANATEL_SERVMOVFIX       Telefone - acessos móveis e fixos em serviço
...                 ...                                                ...
8563             PIBCNE  PIB Estadual - indústria - construção - valor ...
8564              PIBCO  PIB Municipal - serviços - comércio - valor ad...
8565             PIBCOE  PIB Estadual - serviços - comércio - valor adi...
8566               PIBE                   PIB Estadual a preços constantes
8567            PIBEGAE  PIB Estadual - indústria - serviços de utilida...
8568               PIBF  PIB Municipal - serviços - intermediação finan...
8569              PIBFE  PIB Estadual - serviços - intermediação financ...
8570               PIBG  PIB Municipal - serviços - administração públi...
8571              PIBGE  PIB Estadual - serviços - administração públic...
8572               PIBI  PIB Municipal - indústria - valor adicionado -...
8573              PIBIE  PIB Estadual - indústria - valor adicionado - ...
8574             PIBIEE  PIB Estadual - indústria - extrativa mineral -...
8575              PIBIN  PIB Municipal - indústria - transformação e ex...
8576             PIBITE  PIB Estadual - indústria - transformação - val...
8577              PIBOS  PIB Municipal - serviços - outros - valor adic...
8578             PIBOSE  PIB Estadual - serviços - outros - valor adici...
8579             PIBPCE                            PIB Estadual per capita
8580            PIBPMCE         PIB Estadual a preços de mercado  corrente
8581             PIBPOP  População preparada para o PIB Estadual(ponder...
8582             PIBSAE  PIB Estadual - serviços - outros - saúde e edu...
8583             PIBSCE  PIB Estadual - serviços - outros - outros serv...
8584             PIBSDE  PIB Estadual - serv

Let's suppose we are looking for a time series about GDP (PIB, in portuguese). We can filter the results of list_series() function by including a **keyword** to the parameters. This keyword will filter the results for only time series containing the specified keyword in their names:

In [4]:
idpy.list_series("PIB")

CODE                                               NAME
3747   WDI_PPCTAXAPIB    Taxa de paridade do poder de compra (PPC) - PIB
3748      BM_PIBUSDCM                                 PIB - câmbio médio
3749     BM_PIBUSDPPC  PIB - conceito: paridade do poder de compra (PPC)
3750       BM12_PIB12                                                PIB
3751     BM12_PIBAC12                                     PIB - 12 meses
3870        HIST_PIBR                                                PIB
3875       HIST_VAAGR      PIB - agricultura - índice (média 1939 = 100)
3876      HIST_VAAGRL       PIB - agricultura - valor adicionado líquido
3884       HIST_VACOM         PIB - comércio - índice (média 1939 = 100)
3891       HIST_VAGOV          PIB - governo - índice (média 1939 = 100)
3892      HIST_VAGOVL           PIB - governo - valor adicionado líquido
3895      HIST_VAINDL         PIB - indústria - valor adicionado líquido
3896      HIST_VAINDU        PIB - indústria - índice (média 1939 = 100)
3909      HIST_VATOTL                     PIB - valor adicionado líquido
3910   HIST_VATRANSCO  PIB - transportes e comunicações - índice (méd...
3911  HIST_VATRANSCOL  PIB - transportes e comunicações - valor adici...
3912        HIST_YAGG               PIB - agropecuária - var. real anual
3913       HIST_YINDG                  PIB - indústria - var. real anual
3914       HIST_YSERG                   PIB - serviços - var. real anual
3935      SCN_PIBPBN1  PIB - imputação de serviços intermediação fina...
3936    SCN10_AADMY10  PIB - serviços - administração, saúde e educaç...
3957   SCN10_DIPIBG10  PIB - preços de mercado - deflator implícito -...
3985      SCN10_PIB10  PIB - preços de mercado - índice (base 2010 = ...
3986     SCN10_PIBG10          PIB - preços de mercado - var. real anual
3987     SCN10_PIBN10                            PIB - preços de mercado
3988     SCN10_PIBP10              PIB - preços de mercado (preços 2010)
3989    SCN10_PIBPB10    PIB - preços básicos - índice (base 2010 = 100)
3990   SCN10_PIBPBG10             PIB - preços básicos - var. real anual
3991   SCN10_PIBPBN10                               PIB - preços básicos
3992   SCN10_PIBPBP10                 PIB - preços básicos (preços 2010)
...               ...                                                ...
8562           PIBCME  PIB Estadual - serviços - transportes e comuni...
8563           PIBCNE  PIB Estadual - indústria - construção - valor ...
8564            PIBCO  PIB Municipal - serviços - comércio - valor ad...
8565           PIBCOE  PIB Estadual - serviços - comércio - valor adi...
8566             PIBE                   PIB Estadual a preços constantes
8567          PIBEGAE  PIB Estadual - indústria - serviços de utilida...
8568             PIBF  PIB Municipal - serviços - intermediação finan...
8569            PIBFE  PIB Estadual - serviços - intermediação financ...
8570             PIBG  PIB Municipal - serviços - administração públi...
8571            PIBGE  PIB Estadual - serviços - administração públic...
8572             PIBI  PIB Municipal - indústria - valor adicionado -...
8573            PIBIE  PIB Estadual - indústria - valor adicionado - ...
8574           PIBIEE  PIB Estadual - indústria - extrativa mineral -...
8575            PIBIN  PIB Municipal - indústria - transformação e ex...
8576           PIBITE  PIB Estadual - indústria - transformação - val...
8577            PIBOS  PIB Municipal - serviços - outros - valor adic...
8578           PIBOSE  PIB Estadual - serviços - outros - valor adici...
8579           PIBPCE                            PIB Estadual per capita
8580          PIBPMCE         PIB Estadual a preços de mercado  corrente
8581           PIBPOP  População preparada para o PIB Estadual(ponder...
8582           PIBSAE  PIB Estadual - serviços - outros - saúde e edu...
8583           PIBSCE  PIB Estadual - serviços - outros - outros serv...
8584           PIBSDE  PIB Estadual - serviços - outros - serviços do...
85

***Observation:*** Pay attention to the case sensitiveness of the **keyword**. Uppercase/lowercase letters matters!

If you have a more specific need, you can use the parameters **code** or **name**. These parameters will filter the return to only exact matches with the specified **code**, **name** or both at the same time. For example, let's suppose that you have a time series name, say, *"Produção - calcário agrícola - qde."*, but you don't know his code. You can simply solve this problem doing:

In [5]:
idpy.list_series(name="Produção - calcário agrícola - qde.")

CODE                                 NAME
24  ABRACAL_PCALCARIO  Produção - calcário agrícola - qde.

Similarly, if you have a code, say, *GM366_ERC366*, and want to find the name of the time series associated to this code, you can simply do: 

In [6]:
idpy.list_series(code="GM366_ERC366")

CODE                                               NAME
3732  GM366_ERC366  Taxa de câmbio - R$ / US$ - comercial - compra...

You can also use both parameters together, despite the fact that this kind of usage seems to be useless (maybe it will be useful in the case of the existence of two different time series with the same name). 


## 1.1. Having an Overview of a Time Series

Now that you know how to find your desired time series and already have his code, maybe you have an interest in having an overview of his specific informations. For this purpose, there is a function called **describe()**. Let's have a look on *GM366_ERC366* time series:

In [7]:
idpy.describe("GM366_ERC366")

Taxa de câmbio - R$ / US$ - comercial - compra - média
Name            Taxa de câmbio - R$ / US$ - comercial - compra...    
Code                                                 GM366_ERC366    
Big Theme                                          Macroeconômico    
Theme                                                      Câmbio    
Source          Banco Central do Brasil, Sistema Gerenciador d...    
Source acronym                                   Bacen Outras/SGS    
Comment         Taxa de câmbio R$/US$ comercial (valor de comp...    
Last update                         2019-05-15T15:20:01.073-03:00    
Frequency                                                  Diária    
Measure                                                        R$    
Unit                                                         None    
Status                                                          A

## 1.2. Checking Time Series' Data

To access the data of a specific time series we use the function called **timeseries()**. For the full set of data, you just need to specify the desired time series' code:

In [8]:
idpy.timeseries("GM366_ERC366")

YEAR  DAY  MONTH          CODE                       DATE    VALUE (R$)
0      1985    2      1  GM366_ERC366  1985-01-02T00:00:00-02:00  1.152000e-09
1      1985    3      1  GM366_ERC366  1985-01-03T00:00:00-02:00  1.152000e-09
2      1985    4      1  GM366_ERC366  1985-01-04T00:00:00-02:00  1.152000e-09
3      1985    5      1  GM366_ERC366  1985-01-05T00:00:00-02:00           NaN
4      1985    6      1  GM366_ERC366  1985-01-06T00:00:00-02:00           NaN
5      1985    7      1  GM366_ERC366  1985-01-07T00:00:00-02:00  1.173818e-09
6      1985    8      1  GM366_ERC366  1985-01-08T00:00:00-02:00  1.173818e-09
7      1985    9      1  GM366_ERC366  1985-01-09T00:00:00-02:00  1.173818e-09
8      1985   10      1  GM366_ERC366  1985-01-10T00:00:00-02:00  1.173818e-09
9      1985   11      1  GM366_ERC366  1985-01-11T00:00:00-02:00  1.200364e-09
10     1985   12      1  GM366_ERC366  1985-01-12T00:00:00-02:00           NaN
11     1985   13      1  GM366_ERC366  1985-01-13T00:00:00-02:00           NaN
12     1985   14      1  GM366_ERC366  1985-01-14T00:00:00-02:00  1.200364e-09
13     1985   15      1  GM366_ERC366  1985-01-15T00:00:00-02:00  1.200364e-09
14     1985   16      1  GM366_ERC366  1985-01-16T00:00:00-02:00  1.227636e-09
15     1985   17      1  GM366_ERC366  1985-01-17T00:00:00-02:00  1.227636e-09
16     1985   18      1  GM366_ERC366  1985-01-18T00:00:00-02:00  1.227636e-09
17     1985   19      1  GM366_ERC366  1985-01-19T00:00:00-02:00           NaN
18     1985   20      1  GM366_ERC366  1985-01-20T00:00:00-02:00           NaN
19     1985   21      1  GM366_ERC366  1985-01-21T00:00:00-02:00  1.227636e-09
20     1985   22      1  GM366_ERC366  1985-01-22T00:00:00-02:00  1.251273e-09
21     1985   23      1  GM366_ERC366  1985-01-23T00:00:00-02:00  1.251273e-09
22     1985   24      1  GM366_ERC366  1985-01-24T00:00:00-02:00  1.251273e-09
23     1985   25      1  GM366_ERC366  1985-01-25T00:00:00-02:00  1.251273e-09
24     1985   26      1  GM366_ERC366  1985-01-26T00:00:00-02:00           NaN
25     1985   27      1  GM366_ERC366  1985-01-27T00:00:00-02:00           NaN
26     1985   28      1  GM366_ERC366  1985-01-28T00:00:00-02:00  1.270545e-09
27     1985   29      1  GM366_ERC366  1985-01-29T00:00:00-02:00  1.270545e-09
28     1985   30      1  GM366_ERC366  1985-01-30T00:00:00-02:00  1.270545e-09
29     1985   31      1  GM366_ERC366  1985-01-31T00:00:00-02:00  1.297091e-09
...     ...  ...    ...           ...                        ...           ...
12140  2019    2      4  GM366_ERC366  2019-04-02T00:00:00-03:00  3.865500e+00
12141  2019    3      4  GM366_ERC366  2019-04-03T00:00:00-03:00  3.843000e+00
12142  2019    4      4  GM366_ERC366  2019-04-04T00:00:00-03:00  3.870700e+00
12143  2019    5      4  GM366_ERC366  2019-04-05T00:00:00-03:00  3.861600e+00
12144  2019    8      4  GM366_ERC366  2019-04-08T00:00:00-03:00  3.865200e+00
12145  2019    9      4  GM366_ERC366  2019-04-09T00:00:00-03:00  3.855700e+00
12146  2019   10      4  GM366_ERC366  2019-04-10T00:00:00-03:00  3.833900e+00
12147  2019   11      4  GM366_ERC366  2019-04-11T00:00:00-03:00  3.839300e+00
12148  2019   12      4  GM366_ERC366  2019-04-12T00:00:00-03:00  3.867900e+00
12149  2019   15      4  GM366_ERC366  2019-04-15T00:00:00-03:00  3.872400e+00
12150  2019   16      4  GM366_ERC366  2019-04-16T00:00:00-03:00  3.890700e+00
12151  2019   17      4  GM366_ERC366  2019-04-17T00:00:00-03:00  3.921900e+00
12152  2019   18      4  GM366_ERC366  2019-04-18T00:00:00-03:00  3.936400e+00
12153  2019   22      4  GM366_ERC366  2019-04-22T00:00:00-03:00  3.922400e+00
12154  2019   23      4  GM366_ERC366  2019-04-23T00:00:00-03:00  3.943000e+00
12155  2019   24      4  GM366_ERC366  2019-04-24T00:00:00-03:00  3.962400e+00
12156  2019   25      4  GM366_ERC366  2019-04-25T00:00:00-03:00  3.971900e+00
12157  2019   26      4  GM366_ERC366  2019-04-26T00:00:00-03:00  3.934700e+00
12158  2019   29      4  GM366_ERC366  2019-04-29T00:00:0

If you just want the data for a specific year you can use the parameter **year**:

In [9]:
idpy.timeseries("GM366_ERC366", year=2019)

YEAR  DAY  MONTH          CODE                       DATE  VALUE (R$)
12078  2019    2      1  GM366_ERC366  2019-01-02T00:00:00-02:00      3.8589
12079  2019    3      1  GM366_ERC366  2019-01-03T00:00:00-02:00      3.7677
12080  2019    4      1  GM366_ERC366  2019-01-04T00:00:00-02:00      3.7621
12081  2019    7      1  GM366_ERC366  2019-01-07T00:00:00-02:00      3.7056
12082  2019    8      1  GM366_ERC366  2019-01-08T00:00:00-02:00      3.7202
12083  2019    9      1  GM366_ERC366  2019-01-09T00:00:00-02:00      3.6925
12084  2019   10      1  GM366_ERC366  2019-01-10T00:00:00-02:00      3.6863
12085  2019   11      1  GM366_ERC366  2019-01-11T00:00:00-02:00      3.7135
12086  2019   14      1  GM366_ERC366  2019-01-14T00:00:00-02:00      3.7255
12087  2019   15      1  GM366_ERC366  2019-01-15T00:00:00-02:00      3.7043
12088  2019   16      1  GM366_ERC366  2019-01-16T00:00:00-02:00      3.7191
12089  2019   17      1  GM366_ERC366  2019-01-17T00:00:00-02:00      3.7585
12090  2019   18      1  GM366_ERC366  2019-01-18T00:00:00-02:00      3.7480
12091  2019   21      1  GM366_ERC366  2019-01-21T00:00:00-02:00      3.7699
12092  2019   22      1  GM366_ERC366  2019-01-22T00:00:00-02:00      3.7609
12093  2019   23      1  GM366_ERC366  2019-01-23T00:00:00-02:00      3.7988
12094  2019   24      1  GM366_ERC366  2019-01-24T00:00:00-02:00      3.7809
12095  2019   25      1  GM366_ERC366  2019-01-25T00:00:00-02:00      3.7613
12096  2019   28      1  GM366_ERC366  2019-01-28T00:00:00-02:00      3.7670
12097  2019   29      1  GM366_ERC366  2019-01-29T00:00:00-02:00      3.7364
12098  2019   30      1  GM366_ERC366  2019-01-30T00:00:00-02:00      3.7145
12099  2019   31      1  GM366_ERC366  2019-01-31T00:00:00-02:00      3.6513
12100  2019    1      2  GM366_ERC366  2019-02-01T00:00:00-02:00      3.6688
12101  2019    4      2  GM366_ERC366  2019-02-04T00:00:00-02:00      3.6750
12102  2019    5      2  GM366_ERC366  2019-02-05T00:00:00-02:00      3.6735
12103  2019    6      2  GM366_ERC366  2019-02-06T00:00:00-02:00      3.7013
12104  2019    7      2  GM366_ERC366  2019-02-07T00:00:00-02:00      3.7187
12105  2019    8      2  GM366_ERC366  2019-02-08T00:00:00-02:00      3.7178
12106  2019   11      2  GM366_ERC366  2019-02-11T00:00:00-02:00      3.7385
12107  2019   12      2  GM366_ERC366  2019-02-12T00:00:00-02:00      3.7290
...     ...  ...    ...           ...                        ...         ...
12140  2019    2      4  GM366_ERC366  2019-04-02T00:00:00-03:00      3.8655
12141  2019    3      4  GM366_ERC366  2019-04-03T00:00:00-03:00      3.8430
12142  2019    4      4  GM366_ERC366  2019-04-04T00:00:00-03:00      3.8707
12143  2019    5      4  GM366_ERC366  2019-04-05T00:00:00-03:00      3.8616
12144  2019    8      4  GM366_ERC366  2019-04-08T00:00:00-03:00      3.8652
12145  2019    9      4  GM366_ERC366  2019-04-09T00:00:00-03:00      3.8557
12146  2019   10      4  GM366_ERC366  2019-04-10T00:00:00-03:00      3.8339
12147  2019   11      4  GM366_ERC366  2019-04-11T00:00:00-03:00      3.8393
12148  2019   12      4  GM366_ERC366  2019-04-12T00:00:00-03:00      3.8679
12149  2019   15      4  GM366_ERC366  2019-04-15T00:00:00-03:00      3.8724
12150  2019   16      4  GM366_ERC366  2019-04-16T00:00:00-03:00      3.8907
12151  2019   17      4  GM366_ERC366  2019-04-17T00:00:00-03:00      3.9219
12152  2019   18      4  GM366_ERC366  2019-04-18T00:00:00-03:00      3.9364
12153  2019   22      4  GM366_ERC366  2019-04-22T00:00:00-03:00      3.9224
12154  2019   23      4  GM366_ERC366  2019-04-23T00:00:00-03:00      3.9430
12155  2019   24      4  GM366_ERC366  2019-04-24T00:00:00-03:00      3.9624
12156  2019   25      4  GM366_ERC366  2019-04-25T00:00:00-03:00      3.9719
12157  2019   26      4  GM366_ERC366  2019-04-26T00:00:00-03:00      3.9347
12158  2019   29      4  GM366_ERC366  2019-04-29T00:00:00-03:00      3.9358
12159  2019   30      4  GM366_ERC366  2019-04-30T00:00:00-03:00      3.9447
121

If you just want the data for a specific month of a specific year, use both the parameters **year** and **month**:

In [10]:
idpy.timeseries("GM366_ERC366", year=2019, month=4)

YEAR  DAY  MONTH          CODE                       DATE  VALUE (R$)
12139  2019    1      4  GM366_ERC366  2019-04-01T00:00:00-03:00      3.8676
12140  2019    2      4  GM366_ERC366  2019-04-02T00:00:00-03:00      3.8655
12141  2019    3      4  GM366_ERC366  2019-04-03T00:00:00-03:00      3.8430
12142  2019    4      4  GM366_ERC366  2019-04-04T00:00:00-03:00      3.8707
12143  2019    5      4  GM366_ERC366  2019-04-05T00:00:00-03:00      3.8616
12144  2019    8      4  GM366_ERC366  2019-04-08T00:00:00-03:00      3.8652
12145  2019    9      4  GM366_ERC366  2019-04-09T00:00:00-03:00      3.8557
12146  2019   10      4  GM366_ERC366  2019-04-10T00:00:00-03:00      3.8339
12147  2019   11      4  GM366_ERC366  2019-04-11T00:00:00-03:00      3.8393
12148  2019   12      4  GM366_ERC366  2019-04-12T00:00:00-03:00      3.8679
12149  2019   15      4  GM366_ERC366  2019-04-15T00:00:00-03:00      3.8724
12150  2019   16      4  GM366_ERC366  2019-04-16T00:00:00-03:00      3.8907
12151  2019   17      4  GM366_ERC366  2019-04-17T00:00:00-03:00      3.9219
12152  2019   18      4  GM366_ERC366  2019-04-18T00:00:00-03:00      3.9364
12153  2019   22      4  GM366_ERC366  2019-04-22T00:00:00-03:00      3.9224
12154  2019   23      4  GM366_ERC366  2019-04-23T00:00:00-03:00      3.9430
12155  2019   24      4  GM366_ERC366  2019-04-24T00:00:00-03:00      3.9624
12156  2019   25      4  GM366_ERC366  2019-04-25T00:00:00-03:00      3.9719
12157  2019   26      4  GM366_ERC366  2019-04-26T00:00:00-03:00      3.9347
12158  2019   29      4  GM366_ERC366  2019-04-29T00:00:00-03:00      3.9358
12159  2019   30      4  GM366_ERC366  2019-04-30T00:00:00-03:00      3.9447

In [11]:
idpy.timeseries("GM366_ERC366", year=2019, month=4).plot("DAY", "VALUE (R$)")

Similarly, if you just want the data for a specific day of a specific month of a specific year use together the parameters **year**, **month** and **day**:

In [12]:
idpy.timeseries("GM366_ERC366", year=2019, month=4, day=1)

YEAR  DAY  MONTH          CODE                       DATE  VALUE (R$)
12139  2019    1      4  GM366_ERC366  2019-04-01T00:00:00-03:00      3.8676

Another example: returning only data relative to years greater than 2017:

In [13]:
idpy.timeseries("GM366_ERC366", yearGreaterThan=2013)

YEAR  DAY  MONTH          CODE                       DATE  VALUE (R$)
10591  2014    1      1  GM366_ERC366  2014-01-01T00:00:00-02:00         NaN
10592  2014    2      1  GM366_ERC366  2014-01-02T00:00:00-02:00      2.3969
10593  2014    3      1  GM366_ERC366  2014-01-03T00:00:00-02:00      2.3734
10594  2014    4      1  GM366_ERC366  2014-01-04T00:00:00-02:00         NaN
10595  2014    5      1  GM366_ERC366  2014-01-05T00:00:00-02:00         NaN
10596  2014    6      1  GM366_ERC366  2014-01-06T00:00:00-02:00      2.3783
10597  2014    7      1  GM366_ERC366  2014-01-07T00:00:00-02:00      2.3628
10598  2014    8      1  GM366_ERC366  2014-01-08T00:00:00-02:00      2.3773
10599  2014    9      1  GM366_ERC366  2014-01-09T00:00:00-02:00      2.3954
10600  2014   10      1  GM366_ERC366  2014-01-10T00:00:00-02:00      2.3813
10601  2014   11      1  GM366_ERC366  2014-01-11T00:00:00-02:00         NaN
10602  2014   12      1  GM366_ERC366  2014-01-12T00:00:00-02:00         NaN
10603  2014   13      1  GM366_ERC366  2014-01-13T00:00:00-02:00      2.3485
10604  2014   14      1  GM366_ERC366  2014-01-14T00:00:00-02:00      2.3611
10605  2014   15      1  GM366_ERC366  2014-01-15T00:00:00-02:00      2.3465
10606  2014   16      1  GM366_ERC366  2014-01-16T00:00:00-02:00      2.3677
10607  2014   17      1  GM366_ERC366  2014-01-17T00:00:00-02:00      2.3595
10608  2014   18      1  GM366_ERC366  2014-01-18T00:00:00-02:00         NaN
10609  2014   19      1  GM366_ERC366  2014-01-19T00:00:00-02:00         NaN
10610  2014   20      1  GM366_ERC366  2014-01-20T00:00:00-02:00      2.3329
10611  2014   21      1  GM366_ERC366  2014-01-21T00:00:00-02:00      2.3602
10612  2014   22      1  GM366_ERC366  2014-01-22T00:00:00-02:00      2.3595
10613  2014   23      1  GM366_ERC366  2014-01-23T00:00:00-02:00      2.3755
10614  2014   24      1  GM366_ERC366  2014-01-24T00:00:00-02:00      2.4123
10615  2014   25      1  GM366_ERC366  2014-01-25T00:00:00-02:00         NaN
10616  2014   26      1  GM366_ERC366  2014-01-26T00:00:00-02:00         NaN
10617  2014   27      1  GM366_ERC366  2014-01-27T00:00:00-02:00      2.4038
10618  2014   28      1  GM366_ERC366  2014-01-28T00:00:00-02:00      2.4159
10619  2014   29      1  GM366_ERC366  2014-01-29T00:00:00-02:00      2.4390
10620  2014   30      1  GM366_ERC366  2014-01-30T00:00:00-02:00      2.4218
...     ...  ...    ...           ...                        ...         ...
12140  2019    2      4  GM366_ERC366  2019-04-02T00:00:00-03:00      3.8655
12141  2019    3      4  GM366_ERC366  2019-04-03T00:00:00-03:00      3.8430
12142  2019    4      4  GM366_ERC366  2019-04-04T00:00:00-03:00      3.8707
12143  2019    5      4  GM366_ERC366  2019-04-05T00:00:00-03:00      3.8616
12144  2019    8      4  GM366_ERC366  2019-04-08T00:00:00-03:00      3.8652
12145  2019    9      4  GM366_ERC366  2019-04-09T00:00:00-03:00      3.8557
12146  2019   10      4  GM366_ERC366  2019-04-10T00:00:00-03:00      3.8339
12147  2019   11      4  GM366_ERC366  2019-04-11T00:00:00-03:00      3.8393
12148  2019   12      4  GM366_ERC366  2019-04-12T00:00:00-03:00      3.8679
12149  2019   15      4  GM366_ERC366  2019-04-15T00:00:00-03:00      3.8724
12150  2019   16      4  GM366_ERC366  2019-04-16T00:00:00-03:00      3.8907
12151  2019   17      4  GM366_ERC366  2019-04-17T00:00:00-03:00      3.9219
12152  2019   18      4  GM366_ERC366  2019-04-18T00:00:00-03:00      3.9364
12153  2019   22      4  GM366_ERC366  2019-04-22T00:00:00-03:00      3.9224
12154  2019   23      4  GM366_ERC366  2019-04-23T00:00:00-03:00      3.9430
12155  2019   24      4  GM366_ERC366  2019-04-24T00:00:00-03:00      3.9624
12156  2019   25      4  GM366_ERC366  2019-04-25T00:00:00-03:00      3.9719
12157  2019   26      4  GM366_ERC366  2019-04-26T00:00:00-03:00      3.9347
12158  2019   29      4  GM366_ERC366  2019-04-29T00:00:00-03:00      3.9358
12159  2019   30      4  GM366_ERC366  2019-04-30T00:00:00-03:00      3.9447
121

or an interval of years, say, from 2017 to 2018:

In [14]:
idpy.timeseries("GM366_ERC366", yearGreaterThan=2016, yearSmallerThan=2019)

YEAR  DAY  MONTH          CODE                       DATE  VALUE (R$)
11579  2017    2      1  GM366_ERC366  2017-01-02T00:00:00-02:00      3.2723
11580  2017    3      1  GM366_ERC366  2017-01-03T00:00:00-02:00      3.2626
11581  2017    4      1  GM366_ERC366  2017-01-04T00:00:00-02:00      3.2327
11582  2017    5      1  GM366_ERC366  2017-01-05T00:00:00-02:00      3.2123
11583  2017    6      1  GM366_ERC366  2017-01-06T00:00:00-02:00      3.2051
11584  2017    9      1  GM366_ERC366  2017-01-09T00:00:00-02:00      3.2091
11585  2017   10      1  GM366_ERC366  2017-01-10T00:00:00-02:00      3.1912
11586  2017   11      1  GM366_ERC366  2017-01-11T00:00:00-02:00      3.2148
11587  2017   12      1  GM366_ERC366  2017-01-12T00:00:00-02:00      3.1655
11588  2017   13      1  GM366_ERC366  2017-01-13T00:00:00-02:00      3.2028
11589  2017   16      1  GM366_ERC366  2017-01-16T00:00:00-02:00      3.2228
11590  2017   17      1  GM366_ERC366  2017-01-17T00:00:00-02:00      3.2094
11591  2017   18      1  GM366_ERC366  2017-01-18T00:00:00-02:00      3.2205
11592  2017   19      1  GM366_ERC366  2017-01-19T00:00:00-02:00      3.2107
11593  2017   20      1  GM366_ERC366  2017-01-20T00:00:00-02:00      3.1912
11594  2017   23      1  GM366_ERC366  2017-01-23T00:00:00-02:00      3.1603
11595  2017   24      1  GM366_ERC366  2017-01-24T00:00:00-02:00      3.1643
11596  2017   25      1  GM366_ERC366  2017-01-25T00:00:00-02:00      3.1679
11597  2017   26      1  GM366_ERC366  2017-01-26T00:00:00-02:00      3.1798
11598  2017   27      1  GM366_ERC366  2017-01-27T00:00:00-02:00      3.1590
11599  2017   30      1  GM366_ERC366  2017-01-30T00:00:00-02:00      3.1310
11600  2017   31      1  GM366_ERC366  2017-01-31T00:00:00-02:00      3.1264
11601  2017    1      2  GM366_ERC366  2017-02-01T00:00:00-02:00      3.1473
11602  2017    2      2  GM366_ERC366  2017-02-02T00:00:00-02:00      3.1190
11603  2017    3      2  GM366_ERC366  2017-02-03T00:00:00-02:00      3.1235
11604  2017    6      2  GM366_ERC366  2017-02-06T00:00:00-02:00      3.1173
11605  2017    7      2  GM366_ERC366  2017-02-07T00:00:00-02:00      3.1298
11606  2017    8      2  GM366_ERC366  2017-02-08T00:00:00-02:00      3.1248
11607  2017    9      2  GM366_ERC366  2017-02-09T00:00:00-02:00      3.1179
11608  2017   10      2  GM366_ERC366  2017-02-10T00:00:00-02:00      3.1149
...     ...  ...    ...           ...                        ...         ...
12048  2018   19     11  GM366_ERC366  2018-11-19T00:00:00-02:00      3.7547
12049  2018   20     11  GM366_ERC366  2018-11-20T00:00:00-02:00      3.7574
12050  2018   21     11  GM366_ERC366  2018-11-21T00:00:00-02:00      3.7866
12051  2018   22     11  GM366_ERC366  2018-11-22T00:00:00-02:00      3.8097
12052  2018   23     11  GM366_ERC366  2018-11-23T00:00:00-02:00      3.8075
12053  2018   26     11  GM366_ERC366  2018-11-26T00:00:00-02:00      3.8649
12054  2018   27     11  GM366_ERC366  2018-11-27T00:00:00-02:00      3.8919
12055  2018   28     11  GM366_ERC366  2018-11-28T00:00:00-02:00      3.8625
12056  2018   29     11  GM366_ERC366  2018-11-29T00:00:00-02:00      3.8562
12057  2018   30     11  GM366_ERC366  2018-11-30T00:00:00-02:00      3.8627
12058  2018    3     12  GM366_ERC366  2018-12-03T00:00:00-02:00      3.8279
12059  2018    4     12  GM366_ERC366  2018-12-04T00:00:00-02:00      3.8307
12060  2018    5     12  GM366_ERC366  2018-12-05T00:00:00-02:00      3.8555
12061  2018    6     12  GM366_ERC366  2018-12-06T00:00:00-02:00      3.9172
12062  2018    7     12  GM366_ERC366  2018-12-07T00:00:00-02:00      3.8958
12063  2018   10     12  GM366_ERC366  2018-12-10T00:00:00-02:00      3.9104
12064  2018   11     12  GM366_ERC366  2018-12-11T00:00:00-02:00      3.9007
12065  2018   12     12  GM366_ERC366  2018-12-12T00:00:00-02:00      3.8623
12066  2018   13     12  GM366_ERC366  2018-12-13T00:00:00-02:00      3.8784
12067  2018   14     12  GM366_ERC366  2018-12-14T00:00:00-02:00      3.9084
120

or to an interval of months (e.g.: from june to december) for a specifc year, say, 2018:

In [15]:
idpy.timeseries("GM366_ERC366", year=2018, monthGreaterThan=5, monthSmallerThan=13)

YEAR  DAY  MONTH          CODE                       DATE  VALUE (R$)
11931  2018    1      6  GM366_ERC366  2018-06-01T00:00:00-03:00      3.7407
11932  2018    4      6  GM366_ERC366  2018-06-04T00:00:00-03:00      3.7418
11933  2018    5      6  GM366_ERC366  2018-06-05T00:00:00-03:00      3.7746
11934  2018    6      6  GM366_ERC366  2018-06-06T00:00:00-03:00      3.8187
11935  2018    7      6  GM366_ERC366  2018-06-07T00:00:00-03:00      3.8994
11936  2018    8      6  GM366_ERC366  2018-06-08T00:00:00-03:00      3.7853
11937  2018   11      6  GM366_ERC366  2018-06-11T00:00:00-03:00      3.6907
11938  2018   12      6  GM366_ERC366  2018-06-12T00:00:00-03:00      3.7038
11939  2018   13      6  GM366_ERC366  2018-06-13T00:00:00-03:00      3.7048
11940  2018   14      6  GM366_ERC366  2018-06-14T00:00:00-03:00      3.7051
11941  2018   15      6  GM366_ERC366  2018-06-15T00:00:00-03:00      3.7732
11942  2018   18      6  GM366_ERC366  2018-06-18T00:00:00-03:00      3.7537
11943  2018   19      6  GM366_ERC366  2018-06-19T00:00:00-03:00      3.7560
11944  2018   20      6  GM366_ERC366  2018-06-20T00:00:00-03:00      3.7329
11945  2018   21      6  GM366_ERC366  2018-06-21T00:00:00-03:00      3.7888
11946  2018   22      6  GM366_ERC366  2018-06-22T00:00:00-03:00      3.7657
11947  2018   25      6  GM366_ERC366  2018-06-25T00:00:00-03:00      3.7754
11948  2018   26      6  GM366_ERC366  2018-06-26T00:00:00-03:00      3.7715
11949  2018   27      6  GM366_ERC366  2018-06-27T00:00:00-03:00      3.8352
11950  2018   28      6  GM366_ERC366  2018-06-28T00:00:00-03:00      3.8515
11951  2018   29      6  GM366_ERC366  2018-06-29T00:00:00-03:00      3.8552
11952  2018    2      7  GM366_ERC366  2018-07-02T00:00:00-03:00      3.9049
11953  2018    3      7  GM366_ERC366  2018-07-03T00:00:00-03:00      3.8914
11954  2018    4      7  GM366_ERC366  2018-07-04T00:00:00-03:00      3.9052
11955  2018    5      7  GM366_ERC366  2018-07-05T00:00:00-03:00      3.9186
11956  2018    6      7  GM366_ERC366  2018-07-06T00:00:00-03:00      3.9258
11957  2018    9      7  GM366_ERC366  2018-07-09T00:00:00-03:00      3.8680
11958  2018   10      7  GM366_ERC366  2018-07-10T00:00:00-03:00      3.8446
11959  2018   11      7  GM366_ERC366  2018-07-11T00:00:00-03:00      3.8411
11960  2018   12      7  GM366_ERC366  2018-07-12T00:00:00-03:00      3.8558
...     ...  ...    ...           ...                        ...         ...
12048  2018   19     11  GM366_ERC366  2018-11-19T00:00:00-02:00      3.7547
12049  2018   20     11  GM366_ERC366  2018-11-20T00:00:00-02:00      3.7574
12050  2018   21     11  GM366_ERC366  2018-11-21T00:00:00-02:00      3.7866
12051  2018   22     11  GM366_ERC366  2018-11-22T00:00:00-02:00      3.8097
12052  2018   23     11  GM366_ERC366  2018-11-23T00:00:00-02:00      3.8075
12053  2018   26     11  GM366_ERC366  2018-11-26T00:00:00-02:00      3.8649
12054  2018   27     11  GM366_ERC366  2018-11-27T00:00:00-02:00      3.8919
12055  2018   28     11  GM366_ERC366  2018-11-28T00:00:00-02:00      3.8625
12056  2018   29     11  GM366_ERC366  2018-11-29T00:00:00-02:00      3.8562
12057  2018   30     11  GM366_ERC366  2018-11-30T00:00:00-02:00      3.8627
12058  2018    3     12  GM366_ERC366  2018-12-03T00:00:00-02:00      3.8279
12059  2018    4     12  GM366_ERC366  2018-12-04T00:00:00-02:00      3.8307
12060  2018    5     12  GM366_ERC366  2018-12-05T00:00:00-02:00      3.8555
12061  2018    6     12  GM366_ERC366  2018-12-06T00:00:00-02:00      3.9172
12062  2018    7     12  GM366_ERC366  2018-12-07T00:00:00-02:00      3.8958
12063  2018   10     12  GM366_ERC366  2018-12-10T00:00:00-02:00      3.9104
12064  2018   11     12  GM366_ERC366  2018-12-11T00:00:00-02:00      3.9007
12065  2018   12     12  GM366_ERC366  2018-12-12T00:00:00-02:00      3.8623
12066  2018   13     12  GM366_ERC366  2018-12-13T00:00:00-02:00      3.8784
12067  2018   14     12  GM366_ERC366  2018-12-14T00:00:00-02:00      3.9084
120

From now on, use your creativity. There are a lot of possibilities with these parameter combinations. The available parameters for the function **timeseries()** can be found using the function **help()**:

In [16]:
help(idpy.timeseries)

Help on function timeseries in module ipeadatapy.timeseries:

timeseries(series, groupby=None, year=None, yearGreaterThan=None, yearSmallerThan=None, day=None, dayGreaterThan=None, daySmallerThan=None, month=None, monthGreaterThan=None, monthSmallerThan=None, code=None, date=None)
    Returns the specified time series' data values. `series` must be a time series code
    :param series: Time series code. For the available time series run list_series()
    :type series: str
    :param groupby: Grouping criteria
    :type groupby: str, optional
    :param year: Year which the data set will be restricted to.
    :type year: int, optional
    :param yearGreaterThan: Year which the data set will be restricted to years strictly greater.
    :type yearGreaterThan: int, optional
    :param yearSmallerThan: Year which the data set will be restricted to years strictly smaller.
    :type yearSmallerThan: int, optional
    :param day: Day which the data set will be restricted to.
    :type day: int

# 2. INTERMEDIATE: Exploring Ipeadata Metadata

Every Ipeadata's time series is accompanied by a set of metadata. Some examples of the elements of this set of metadatas are: *country*, *big theme*, *theme*, *source*, *unit of measure*, *frequency* et cetera.

Some specific kinds of metadata have their own function on Ipeadata API. For example, you can have a look at the available Ipeadata's countries by running the **countries()** function:

In [17]:
idpy.countries()

ID                         COUNTRY
0    ZAF                   África do Sul
1    DEU                        Alemanha
2   LATI                  América Latina
3    AGO                          Angola
4    SAU                  Arábia Saudita
5    DZA                         Argélia
6    ARG                       Argentina
7    AUS                       Austrália
8    AUT                         Áustria
9    BEL                         Bélgica
10   BOL                         Bolívia
11   BRA                          Brasil
12   CPV                      Cabo Verde
13   CAN                          Canadá
14   CHL                           Chile
15   CHN                           China
16   SGP                       Cingapura
17   COL                        Colômbia
18   COG                           Congo
19   KOR                   Coréia do Sul
20   DNK                       Dinamarca
21   ARE          Emirados Árabes Unidos
22   ECU                         Equador
23   ESP                         Espanha
24   USA                  Estados Unidos
25   RUS                 Federação Russa
26   FIN                       Finlândia
27   FRA                          França
28   GBR  Grã-Bretanha (Reino Unido, UK)
29   GRC                          Grécia
..   ...                             ...
37   IRL                         Irlanda
38   ISL                        Islândia
39   ISR                          Israel
40   ITA                          Itália
41   JPN                           Japão
42   LUX                      Luxemburgo
43   MAC                           Macau
44   MYS                         Malásia
45   MAR                        Marrocos
46   MEX                          Mexico
47   MOZ                      Moçambique
48  WRLD                         Mundial
49   MMR               Myanma (Ex-Burma)
50   NGA                         Nigéria
51   NOR                         Noruega
52   NZL                   Nova Zelândia
53  INDU         Países industrializados
54   PRY                        Paraguai
55   PER                            Peru
56   PRT                        Portugal
57   QAT                           Qatar
58   CZE                República Tcheca
59   STP             São Tomé e Príncipe
60   SWE                          Suécia
61   CHE                           Suíça
62   THA                       Tailândia
63   TLS     Timor Leste (Ex-East Timor)
64   URY                         Uruguai
65   VEN                       Venezuela
66  ZEUR                    Zona do Euro

[67 rows x 2 columns]

Examples of utilities for the function **countries()** are: (...)

You can also have a look on the available themes for Ipeadata using the function **themes()**:

In [18]:
idpy.themes()

ID                     NAME  MACRO  REGIONAL  SOCIAL
0   28             Agropecuária    NaN       1.0     NaN
1   23       Assistência social    NaN       NaN     1.0
2   25     Avaliação do governo    NaN       NaN     NaN
3   10    Balanço de pagamentos    1.0       NaN     NaN
4    7                   Câmbio    1.0       NaN     NaN
5    5        Comércio exterior    1.0       1.0     NaN
6    2         Consumo e vendas    1.0       1.0     NaN
7    8         Contas nacionais    1.0       NaN     NaN
8   81         Contas Regionais    NaN       1.0     NaN
9   24       Correção monetária    1.0       NaN     NaN
10  37               Demografia    NaN       NaN     1.0
11  54        Deputado Estadual    NaN       1.0     NaN
12  55         Deputado Federal    NaN       1.0     NaN
13  38   Desenvolvimento humano    NaN       NaN     1.0
14  11   Economia internacional    1.0       NaN     NaN
15  29                 Educação    NaN       NaN     1.0
16  18                 Eleições    NaN       NaN     NaN
17  63               Eleitorado    NaN       1.0     NaN
18  12                  Emprego    1.0       1.0     NaN
19  19       Estoque de capital    1.0       1.0     NaN
20   6        Finanças públicas    1.0       1.0     NaN
21  39              Financeiras    1.0       NaN     NaN
22  32               Geográfico    NaN       1.0     NaN
23  56               Governador    NaN       1.0     NaN
24  31                Habitação    NaN       NaN     1.0
25  80                 IDHm1991    NaN       NaN     NaN
26  79                 IDHm2000    NaN       NaN     1.0
27  78                 IDHm2010    NaN       NaN     NaN
28  15      Indicadores sociais    1.0       NaN     NaN
29   4                    Juros    NaN       NaN     NaN
30  40      Mercado de trabalho    NaN       NaN     1.0
31   3          Moeda e crédito    1.0       1.0     NaN
32  27  Percepção e expectativa    1.0       NaN     NaN
33  14                População    1.0       1.0     NaN
34   9                   Preços    1.0       1.0     NaN
35  57                 Prefeito    NaN       1.0     NaN
36  58               Presidente    NaN       1.0     NaN
37   1                 Produção    1.0       1.0     NaN
38  16                Projeções    1.0       NaN     NaN
39  30                    Renda    NaN       1.0     1.0
40  13          Salário e renda    1.0       NaN     NaN
41  41                    Saúde    NaN       NaN     1.0
42  20        Segurança Pública    NaN       1.0     NaN
43  59                  Senador    NaN       1.0     NaN
44  17   Sinopse macroeconômica    1.0       NaN     NaN
45  33               Transporte    1.0       1.0     NaN
46  26                   Vendas    NaN       1.0     NaN
47  60                 Vereador    NaN       1.0     NaN

Let's suppose you have the interest to know which of the themes of Ipeadata are related to the Macroeconomics big theme. The parameter **macro** will solve this problem:

In [19]:
idpy.themes(macro=1)

ID                     NAME  MACRO  REGIONAL  SOCIAL
3   10    Balanço de pagamentos    1.0       NaN     NaN
4    7                   Câmbio    1.0       NaN     NaN
5    5        Comércio exterior    1.0       1.0     NaN
6    2         Consumo e vendas    1.0       1.0     NaN
7    8         Contas nacionais    1.0       NaN     NaN
9   24       Correção monetária    1.0       NaN     NaN
14  11   Economia internacional    1.0       NaN     NaN
18  12                  Emprego    1.0       1.0     NaN
19  19       Estoque de capital    1.0       1.0     NaN
20   6        Finanças públicas    1.0       1.0     NaN
21  39              Financeiras    1.0       NaN     NaN
28  15      Indicadores sociais    1.0       NaN     NaN
31   3          Moeda e crédito    1.0       1.0     NaN
32  27  Percepção e expectativa    1.0       NaN     NaN
33  14                População    1.0       1.0     NaN
34   9                   Preços    1.0       1.0     NaN
37   1                 Produção    1.0       1.0     NaN
38  16                Projeções    1.0       NaN     NaN
40  13          Salário e renda    1.0       NaN     NaN
44  17   Sinopse macroeconômica    1.0       NaN     NaN
45  33               Transporte    1.0       1.0     NaN

Let's now suppose that you just want the function to return **themes** that are related both to the macroeconomics and regional themes. For this, use **macro** and **regional** parameters together:

In [20]:
idpy.themes()

ID                     NAME  MACRO  REGIONAL  SOCIAL
0   28             Agropecuária    NaN       1.0     NaN
1   23       Assistência social    NaN       NaN     1.0
2   25     Avaliação do governo    NaN       NaN     NaN
3   10    Balanço de pagamentos    1.0       NaN     NaN
4    7                   Câmbio    1.0       NaN     NaN
5    5        Comércio exterior    1.0       1.0     NaN
6    2         Consumo e vendas    1.0       1.0     NaN
7    8         Contas nacionais    1.0       NaN     NaN
8   81         Contas Regionais    NaN       1.0     NaN
9   24       Correção monetária    1.0       NaN     NaN
10  37               Demografia    NaN       NaN     1.0
11  54        Deputado Estadual    NaN       1.0     NaN
12  55         Deputado Federal    NaN       1.0     NaN
13  38   Desenvolvimento humano    NaN       NaN     1.0
14  11   Economia internacional    1.0       NaN     NaN
15  29                 Educação    NaN       NaN     1.0
16  18                 Eleições    NaN       NaN     NaN
17  63               Eleitorado    NaN       1.0     NaN
18  12                  Emprego    1.0       1.0     NaN
19  19       Estoque de capital    1.0       1.0     NaN
20   6        Finanças públicas    1.0       1.0     NaN
21  39              Financeiras    1.0       NaN     NaN
22  32               Geográfico    NaN       1.0     NaN
23  56               Governador    NaN       1.0     NaN
24  31                Habitação    NaN       NaN     1.0
25  80                 IDHm1991    NaN       NaN     NaN
26  79                 IDHm2000    NaN       NaN     1.0
27  78                 IDHm2010    NaN       NaN     NaN
28  15      Indicadores sociais    1.0       NaN     NaN
29   4                    Juros    NaN       NaN     NaN
30  40      Mercado de trabalho    NaN       NaN     1.0
31   3          Moeda e crédito    1.0       1.0     NaN
32  27  Percepção e expectativa    1.0       NaN     NaN
33  14                População    1.0       1.0     NaN
34   9                   Preços    1.0       1.0     NaN
35  57                 Prefeito    NaN       1.0     NaN
36  58               Presidente    NaN       1.0     NaN
37   1                 Produção    1.0       1.0     NaN
38  16                Projeções    1.0       NaN     NaN
39  30                    Renda    NaN       1.0     1.0
40  13          Salário e renda    1.0       NaN     NaN
41  41                    Saúde    NaN       NaN     1.0
42  20        Segurança Pública    NaN       1.0     NaN
43  59                  Senador    NaN       1.0     NaN
44  17   Sinopse macroeconômica    1.0       NaN     NaN
45  33               Transporte    1.0       1.0     NaN
46  26                   Vendas    NaN       1.0     NaN
47  60                 Vereador    NaN       1.0     NaN

The parameter **social** is also available and works in the same way of **macro** and **regional**. For more parameters available for the function **themes()** run **help(idpy.themes)**

Other important metadata is the source. This metadata have his own functions, **sources()**. Let's have a look:

In [21]:
idpy.sources()

0                         Abia
1                       Abinee
2                         ABPO
3                      Abracal
4                        Abras
5                    ACSP/IEGV
6                         Anac
7                       Anatel
8                       Anbima
9                       Anbima
10                        Anda
11                     Anfavea
12                         ANP
13                   Antaq/AEP
14                   ANTT/AETT
15                 Assembleias
16            Bacen Outras/BPE
17             Bacen Outras/IA
18           Bacen Outras/Iceb
19             Bacen Outras/IE
20            Bacen Outras/ISP
21            Bacen Outras/SGS
22           Bacen Outras/SHFP
23     Bacen/Boletim/Ativ. Ec.
24            Bacen/Boletim/BP
25      Bacen/Boletim/Bpantigo
26       Bacen/Boletim/Intern.
27     Bacen/Boletim/M. Finan.
28             Bacen/BP (BPM5)
29             Bacen/BP (BPM6)
                ...           
128           Min. Fazenda/STN
129     

For regional time series we also have some information about Brazilian territories through the function **territories()**:

In [22]:
idpy.territories()

NAME            ID  \
0                                         (não definido)                 
1                                                 Brasil             0   
2                                           Região Norte             1   
3                                               Rondônia            11   
4                                  Alta Floresta D'Oeste       1100015   
5                                              Ariquemes       1100023   
6                                                 Cabixi       1100031   
7                                                 Cacoal       1100049   
8                                             Cerejeiras       1100056   
9                                      Colorado do Oeste       1100064   
10                                            Corumbiara       1100072   
11                                         Costa Marques       1100080   
12                                       Espigão D'Oeste        110009   
13                                       Espigão D'Oeste       1100098   
14                                           Porto Velho         11001   
15                                         Guajará-Mirim        110010   
16                                         Guajará-Mirim       1100106   
17                                                  Jaru       1100114   
18                                             Ji-Paraná        110012   
19                                             Ji-Paraná       1100122   
20                                    Machadinho D'Oeste       1100130   
21                              Nova Brasilândia D'Oeste        110014   
22                              Nova Brasilândia D'Oeste       1100148   
23                                   Ouro Preto do Oeste       1100155   
24                                         Pimenta Bueno       1100189   
25                                         Guajará-Mirim         11002   
26                                           Porto Velho       1100205   
27                                     Presidente Médici        110025   
28                                     Presidente Médici       1100254   
29                                            Rio Crespo       1100262   
...                                                  ...           ...   
19319                                        AMC9197 007  52AMC9197007   
19320                                        AMC9197 008  52AMC9197008   
19321                                        AMC9197 009  52AMC9197009   
19322                                        AMC9197 010  52AMC9197010   
19323                                        AMC9197 011  52AMC9197011   
19324                                        AMC9197 012  52AMC9197012   
19325                                        AMC9197 013  52AMC9197013   
19326                                        AMC9197 014  52AMC9197014   
19327                                        AMC9197 015  52AMC9197015   
19328                                        AMC9197 016  52AMC9197016   
19329                                        AMC9197 017  52AMC9197017   
19330                                        AMC9197 018  52AMC9197018   
19331                                        AMC9197 019  52AMC9197019   
19332                                        AMC9197 020  52AMC9197020   
19333                                        AMC9197 021  52AMC9197021   
19334                                        AMC9197 022  52AMC9197022   
19335                                        AMC9197 023  52AMC9197023   
19336                                        AMC9197 024  52AMC9197024   
19337                                        AMC9197 025  52AMC9197025   
19338                                   Distrito Federal            53   
19339              Área Metropolitana de Goiânia - Goiás      53 ME 01   
19340  Área Metropolitana de Brasília - Distrito Federal     53 ME 01A   
19341                     Área Metropolitana de Brasília

Two interesting parameters of **territories()** function are **areaGreaterThan** and **areaSmallerThan**. With these parameters, it is possible to filter the return of the function for just territories greater than, smaller than or between the specified parameters. For example, let's check which of the Brazilian territories have the area greater than 1000000:

In [23]:
idpy.territories(areaGreaterThan=1000000)

NAME                  ID        LEVEL       AREA CAPITAL
1                   Brasil                   0       Brasil  8531507.6   False
2             Região Norte                   1      Regiões  3869637.9   False
138               Amazonas                  13      Estados  1577820.2   False
386                   Pará                  15      Estados  1253164.5   False
1161       Região Nordeste                   2      Regiões  1558200.4   False
17960  Região Centro-oeste                   5      Regiões  1612077.2   False
18452     AMC1872_1997 001  513AMC1872_1997001  AMC 1872-00  1947986.1    None
18454          AMC2097 001        51AMC2097001    AMC 20-00  1061175.7    None

Let's now check the territories which the area is between 1000000 and 1100000:

In [24]:
idpy.territories(areaGreaterThan=1000000, areaSmallerThan=1500000)

NAME            ID      LEVEL       AREA CAPITAL
386           Pará            15    Estados  1253164.5   False
18454  AMC2097 001  51AMC2097001  AMC 20-00  1061175.7    None

## 2.1. Advanced Filtering Using Metadata

Now that you have knowledge of some of the metadata of Ipeadata, let's introduce yourself to a function called **metadata()**. This function returns all Ipeadata's time series in a data frame, similarly to the **list_series()** function. However, the difference between the two functions is that **metadata()** returns not only the time series but also their metadata. You might then be asking yourself why these two functions exists, since **metadata()** is a more complete version of the list_series function (**metadata()** features all of the **list_series()** information plus metadata). The answer is: **list_series()** is intended to be a more simplistic version, aiming unexperienced users and designed to be friendly to them. **metadata()**, in fact, is a more complete version as well as more confusing because of the quantity of information returned. No more words, let's run the function:

In [25]:
idpy.metadata()

BIG THEME                                             SOURCE  \
0     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
1     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
2     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
3     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
4     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
5     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
6     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
7     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
8     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
9     Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
10    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
11    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
12    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
13    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
14    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
15    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
16    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
17    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
18    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
19    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
20    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
21    Macroeconômico  Instituto Brasileiro de Geografia e Estatístic...   
22    Macroeconômico  Associação Brasileira das Indústrias da Alimen...   
23    Macroeconômico  Associação Brasileira do Papelão Ondulado, Bol...   
24    Macroeconômico  Associação Brasileira dos Produtores de Calcár...   
25    Macroeconômico  Outras fontes, inclusive compilação de vários ...   
26    Macroeconômico      Agência Nacional de Telecomunicações (Anatel)   
27    Macroeconômico      Agência Nacional de Telecomunicações (Anatel)   
28    Macroeconômico      Agência Nacional de Telecomunicações (Anatel)   
29    Macroeconômico      Agência Nacional de Telecomunicações (Anatel)   
...              ...                                                ...   
8563        Regional   Instituto  Brasileiro de Geografia e Estatística   
8564        Regional   Instituto  Brasileiro de Geografia e Estatística   
8565        Regional   Instituto  Brasileiro de Geografia e Estatística   
8566        Regional   Instituto  Brasileiro de Geografia e Estatística   
8567        Regional   Instituto  Brasileiro de Geografia e Estatística   
8568        Regional   Instituto  Brasileiro de Geografia e Estatística   
8569        Regional   Instituto  Brasileiro de Geografia e Estatística   
8570        Regional   Instituto  Brasileiro de Geografia e Estatística   
8571        Regional   Instituto  Brasileiro de Geografia e Estatística   
8572        Regional   Instituto  Brasileiro de Geografia e Estatística   
8573        Regional   Instituto  Brasileiro de Geografia e Estatística   
8574        Regional   Instituto  Brasileiro de Geografia e Estatística   
8575        Regional   Instituto  Brasileiro de Geografia e Estatística   
8576        Regional   Instituto  Brasileiro de Geografia e Estatística   
8577        Regional   Instituto  Brasileiro de Geografia e Estatística   
8578        Regional   Instituto  Brasileiro de Geografia e Estatística   
8579        Regional   Instituto  Brasileiro de Geografia e Estatística   
8580        Regional   Instituto  Brasileiro de Geografia e Estatística   
8581        Regional   Instituto  Brasileiro de Geografia e Estatística   
8582        Regional   Instituto  Brasileiro de Geografia e Estatística   
8583        Regional   Instituto  Brasileiro de Geografia e Estatística   
8584        Regional   Instituto  Br

But why is this function so powerful and important? The first obvious answer is: it gives you more informations about time series. The not-so-obvious answer is: it allows you to better filter time series from Ipeadata. Let's state an illustrative problem for better understanding:

*Ipeadata API has 8565 time series in total. Let's suppose you are doing research in macroeconomics about the United States, but for some specific reason, your interest in data is restricted to data published by The Economist. It also needs to be quarterly published. How to solve this problem using ipeadatapy Python package?*

In [26]:
idpy.metadata(big_theme="Macroeconômico", country="USA", source="Economist", frequency="Trimestral")

BIG THEME         SOURCE SOURCE ACRONYM         SOURCE URL  \
5629  Macroeconômico  The Economist      Economist  www.economist.com   
5630  Macroeconômico  The Economist      Economist  www.economist.com   
5631  Macroeconômico  The Economist      Economist  www.economist.com   

         UNIT COUNTRY   FREQUENCY                    LAST UPDATE  \
5629  bilhões     USA  Trimestral      2018-12-29T07:52:01-02:00   
5630     None     USA  Trimestral  2019-03-18T13:58:01.433-03:00   
5631     None     USA  Trimestral  2019-03-18T13:58:01.433-03:00   

                  CODE                          COMMENT  \
5629    ECONMI4_USCCY4  Série descontinuada pela fonte.   
5630  ECONMI4_USPIBG34     Produto Interno Bruto (PIB).   
5631   ECONMI4_USPIBG4     Produto Interno Bruto (PIB).   

                                                   NAME  NUMERICA  \
5629  balanço - conta corrente - saldo (acum. 12 meses)      True   
5630                PIB - var. real trimestral anualiz.      True   
5631  PIB - var. real contra igual trimestre do ano ...      True   

     SERIES STATUS  THEME CODE   MEASURE  
5629             I          11       US$  
5630             A          11  (% a.a.)  
5631             A          11  (% a.a.)

*Gotcha!* Other metadata also can be used as filtering parameters. For all parameters run help(idpy.metadata).

In [27]:
help(idpy.metadata)

Help on function metadata in module ipeadatapy.metadata:

metadata(series=None, big_theme=None, source=None, country=None, frequency=None, unit=None, measure=None, status=None, source_ext=None, source_url=None, last_update=None, code=None, comment=None, name=None, numerica=None, theme_id=None)
    :param series: Time series code.
    :type series: str, optional
    :param big_theme: Big theme by which the return will be fitered. Options: "Macroeconômico", "Regional" or "Social"
    :type big_theme: str, optional
    :param source: Source by which the return will be filtered. For available sources run sources() function.
    :type source: str, optional
    :param country: Country ID by which the return will be filtered. For available countries and their IDs run countries() function.
    :type country: str, optional
    :param frequency: Frequency by which the return will be filtered.
    :type frequency: str, optional
    :param unit: Unit by which the return will be filtered.
    :type

# 3. Extracting Data

In this example we will extract GM366_ERC366 time series data for april of 2019 in the format of a csv file (;-separated) named as "FileName.csv" to the same folder where this Python notebook is running.

In [33]:
idpy.timeseries("GM366_ERC366", year=2019, month=4).to_csv("FileName.csv", sep=";")

Another option is to extract the time series as a xlsx file:

In [36]:
idpy.timeseries("GM366_ERC366", year=2019, month=4).to_excel("FileName.xlsx")

# 4. ADVANCED: Doing from Scratch

http://ipeadata2-homologa.ipea.gov.br/api/

Use idpy.api_call(). Example:

In [62]:
idpy.api_call("http://ipeadata2-homologa.ipea.gov.br/api/v1/Paises")

PAICODIGO                         PAINOME
0        ZAF                   África do Sul
1        DEU                        Alemanha
2       LATI                  América Latina
3        AGO                          Angola
4        SAU                  Arábia Saudita
5        DZA                         Argélia
6        ARG                       Argentina
7        AUS                       Austrália
8        AUT                         Áustria
9        BEL                         Bélgica
10       BOL                         Bolívia
11       BRA                          Brasil
12       CPV                      Cabo Verde
13       CAN                          Canadá
14       CHL                           Chile
15       CHN                           China
16       SGP                       Cingapura
17       COL                        Colômbia
18       COG                           Congo
19       KOR                   Coréia do Sul
20       DNK                       Dinamarca
21       ARE          Emirados Árabes Unidos
22       ECU                         Equador
23       ESP                         Espanha
24       USA                  Estados Unidos
25       RUS                 Federação Russa
26       FIN                       Finlândia
27       FRA                          França
28       GBR  Grã-Bretanha (Reino Unido, UK)
29       GRC                          Grécia
..       ...                             ...
37       IRL                         Irlanda
38       ISL                        Islândia
39       ISR                          Israel
40       ITA                          Itália
41       JPN                           Japão
42       LUX                      Luxemburgo
43       MAC                           Macau
44       MYS                         Malásia
45       MAR                        Marrocos
46       MEX                          Mexico
47       MOZ                      Moçambique
48      WRLD                         Mundial
49       MMR               Myanma (Ex-Burma)
50       NGA                         Nigéria
51       NOR                         Noruega
52       NZL                   Nova Zelândia
53      INDU         Países industrializados
54       PRY                        Paraguai
55       PER                            Peru
56       PRT                        Portugal
57       QAT                           Qatar
58       CZE                República Tcheca
59       STP             São Tomé e Príncipe
60       SWE                          Suécia
61       CHE                           Suíça
62       THA                       Tailândia
63       TLS     Timor Leste (Ex-East Timor)
64       URY                         Uruguai
65       VEN                       Venezuela
66      ZEUR                    Zona do Euro

[67 rows x 2 columns]